### ISSUE TITLE
Crash when slicing using a scalar extracted from an array

### ISSUE THREAD

**Issue by manopapad** [nv-legate/cunumeric#472]
Bug originally discovered by @rohany, minimized by yours truly.

The following program:

```
import cunumeric as cn
offsets = cn.ones((1,), dtype=int)  # edit: have to set integer type for this to work
offset = offsets[0]
x = cn.zeros((300,))
x[:offset] = 0
```

causes the following error:

```
Traceback (most recent call last):
  File "/gpfs/fs1/mpapadakis/legate.core/install/lib/python3.8/site-packages/legion_top.py", line 440, in legion_python_main
    run_path(args[start], run_name='__main__')
  File "/gpfs/fs1/mpapadakis/legate.core/install/lib/python3.8/site-packages/legion_top.py", line 255, in run_path
    exec(code, module.__dict__, module.__dict__)
  File "a.py", line 12, in <module>
    x[:offset] = 0
  File "/gpfs/fs1/mpapadakis/cunumeric/cunumeric/coverage.py", line 107, in wrapper
    return func(*args, **kwargs)
  File "/gpfs/fs1/mpapadakis/cunumeric/cunumeric/array.py", line 105, in wrapper
    return func(*args, **kwargs)
  File "/gpfs/fs1/mpapadakis/cunumeric/cunumeric/array.py", line 1513, in __setitem__
    self._thunk.set_item(key, value._thunk)
  File "/gpfs/fs1/mpapadakis/cunumeric/cunumeric/deferred.py", line 116, in wrapper
    return func(*args, **kwargs)
  File "/gpfs/fs1/mpapadakis/cunumeric/cunumeric/deferred.py", line 862, in set_item
    view.copy(rhs, deep=False)
  File "/gpfs/fs1/mpapadakis/cunumeric/cunumeric/deferred.py", line 116, in wrapper
    return func(*args, **kwargs)
  File "/gpfs/fs1/mpapadakis/cunumeric/cunumeric/deferred.py", line 354, in copy
    self.unary_op(
  File "/gpfs/fs1/mpapadakis/cunumeric/cunumeric/deferred.py", line 116, in wrapper
    return func(*args, **kwargs)
  File "/gpfs/fs1/mpapadakis/cunumeric/cunumeric/deferred.py", line 2237, in unary_op
    task.execute()
  File "/gpfs/fs1/mpapadakis/legate.core/install/lib/python3.8/site-packages/legate/core/operation.py", line 199, in execute
    self._context.runtime.submit(self)
  File "/gpfs/fs1/mpapadakis/legate.core/install/lib/python3.8/site-packages/legate/core/runtime.py", line 1052, in submit
    self.flush_scheduling_window()
  File "/gpfs/fs1/mpapadakis/legate.core/install/lib/python3.8/site-packages/legate/core/runtime.py", line 1045, in flush_scheduling_window
    self._schedule(ops)
  File "/gpfs/fs1/mpapadakis/legate.core/install/lib/python3.8/site-packages/legate/core/runtime.py", line 1033, in _schedule
    op.launch(strategy)
  File "/gpfs/fs1/mpapadakis/legate.core/install/lib/python3.8/site-packages/legate/core/operation.py", line 529, in launch
    req, tag, store_part = get_requirement(store, part_symb)
  File "/gpfs/fs1/mpapadakis/legate.core/install/lib/python3.8/site-packages/legate/core/operation.py", line 518, in get_requirement
    req = store_part.get_requirement(strategy.launch_ndim)
  File "/gpfs/fs1/mpapadakis/legate.core/install/lib/python3.8/site-packages/legate/core/store.py", line 857, in get_requirement
    part = self._storage_partition.find_or_create_legion_partition()
  File "/gpfs/fs1/mpapadakis/legate.core/install/lib/python3.8/site-packages/legate/core/store.py", line 515, in find_or_create_legion_partition
    return self._parent.find_or_create_legion_partition(
  File "/gpfs/fs1/mpapadakis/legate.core/install/lib/python3.8/site-packages/legate/core/store.py", line 800, in find_or_create_legion_partition
    assert isinstance(self.data, RegionField)
  File "/gpfs/fs1/mpapadakis/legate.core/install/lib/python3.8/site-packages/legate/core/store.py", line 617, in data
    self._data = self._parent.get_child_data(self._color)
  File "/gpfs/fs1/mpapadakis/legate.core/install/lib/python3.8/site-packages/legate/core/store.py", line 501, in get_child_data
    data = parent_data.get_child(
  File "/gpfs/fs1/mpapadakis/legate.core/install/lib/python3.8/site-packages/legate/core/store.py", line 408, in get_child
    if functor in self._partitions:
  File "/gpfs/fs1/mpapadakis/legate.core/install/lib/python3.8/site-packages/legate/core/partition.py", line 216, in __hash__
    self._hash = hash(
  File "/gpfs/fs1/mpapadakis/legate.core/install/lib/python3.8/site-packages/legate/core/shape.py", line 130, in __hash__
    return hash((self.__class__, True, self._extents))
  File "/gpfs/fs1/mpapadakis/cunumeric/cunumeric/coverage.py", line 107, in wrapper
    return func(*args, **kwargs)
  File "/gpfs/fs1/mpapadakis/cunumeric/cunumeric/array.py", line 875, in __hash__
    raise TypeError("unhashable type: cunumeric.ndarray")
TypeError: unhashable type: cunumeric.ndarray
```

**Comment by manopapad** [comment]
Adding some extra cases that should work out of the box, inspired after discussions with @jjwilke:

```
a = cn.ones((10,), dtype=int)
print(a[0])
b = np.arange(5)
print(b[a[0]])
print(b[:a[0]])
c = [0, 1, 2, 3, 4]
print(c[a[0]])
print(c[:a[0]])
```

**Comment by manopapad** [comment]
Fixed after #479
